File: framework/src/datamirai_engine/memory/in_memory_backend.py
"""InMemoryBackend -- default backend for dev/testing.

Same behavior as original LongTermMemory + SharedLog, wrapped in the
MemoryBackend interface. No persistence -- data lives in RAM only.
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Any

from .backend import LogEntry, MemoryBackend, MemoryEntry


class InMemoryBackend(MemoryBackend):
    """In-memory backend. For dev/testing. Same behavior as current LongTermMemory."""

    def __init__(self) -> None:
        self._memories: list[MemoryEntry] = []
        self._logs: list[LogEntry] = []

    async def save_learning(
        self,
        agent_id: str,
        session_id: str,
        summary: str,
        decisions: list[str] | None = None,
        learnings: list[str] | None = None,
        tags: list[str] | None = None,
    ) -> str:
        memory_id = str(uuid.uuid4())
        entry = MemoryEntry(
            id=memory_id,
            agent_id=agent_id,
            session_id=session_id,
            summary=summary,
            decisions=decisions or [],
            learnings=learnings or [],
            tags=tags or [],
            created_at=datetime.now(timezone.utc).isoformat(),
        )
        self._memories.append(entry)
        return memory_id

    async def search(
        self, agent_id: str, query: str, limit: int = 10
    ) -> list[MemoryEntry]:
        query_lower = query.lower()
        matches: list[MemoryEntry] = []
        for e in self._memories:
            if e.agent_id != agent_id:
                continue
            if (
                query_lower in e.summary.lower()
                or any(query_lower in t.lower() for t in e.tags)
                or any(query_lower in d.lower() for d in e.decisions)
                or any(query_lower in l.lower() for l in e.learnings)
            ):
                matches.append(e)
        # Most recent first
        matches.sort(key=lambda m: m.created_at, reverse=True)
        return matches[:limit]

    async def get_recent(
        self, agent_id: str, limit: int = 10
    ) -> list[MemoryEntry]:
        agent_entries = [e for e in self._memories if e.agent_id == agent_id]
        agent_entries.sort(key=lambda m: m.created_at, reverse=True)
        return agent_entries[:limit]

    async def delete(self, memory_id: str) -> bool:
        for i, e in enumerate(self._memories):
            if e.id == memory_id:
                self._memories.pop(i)
                return True
        return False

    async def save_log(
        self,
        agent_id: str,
        session_id: str,
        message: str,
        metadata: dict[str, Any] | None = None,
    ) -> str:
        log_id = str(uuid.uuid4())
        entry = LogEntry(
            id=log_id,
            agent_id=agent_id,
            session_id=session_id,
            message=message,
            metadata=metadata or {},
            created_at=datetime.now(timezone.utc).isoformat(),
        )
        self._logs.append(entry)
        return log_id

    async def get_logs(
        self,
        agent_id: str,
        limit: int = 50,
        session_id: str | None = None,
    ) -> list[LogEntry]:
        entries = [e for e in self._logs if e.agent_id == agent_id]
        if session_id:
            entries = [e for e in entries if e.session_id == session_id]
        entries.sort(key=lambda e: e.created_at, reverse=True)
        return entries[:limit]

    async def count(self, agent_id: str) -> int:
        return sum(1 for e in self._memories if e.agent_id == agent_id)

File: rust/engine/src/memory/in_memory_backend.rs
//! InMemoryBackend -- default MemoryBackend for dev/testing.
//!
//! No persistence -- data lives in RAM only. Thread-safe via `Arc<RwLock<_>>`.

use std::sync::Arc;

use async_trait::async_trait;
use tokio::sync::RwLock;
use uuid::Uuid;

use crate::core::RunnerError;

use super::long_term::{LongTermEntry, MemoryBackend};

/// In-memory backend backed by a shared `Vec`. Safe to clone and share across tasks.
#[derive(Debug, Clone)]
pub struct InMemoryBackend {
    entries: Arc<RwLock<Vec<LongTermEntry>>>,
}

impl InMemoryBackend {
    /// Create an empty backend.
    pub fn new() -> Self {
        Self {
            entries: Arc::new(RwLock::new(Vec::new())),
        }
    }
}

impl Default for InMemoryBackend {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl MemoryBackend for InMemoryBackend {
    async fn save(&self, entry: &LongTermEntry) -> Result<String, RunnerError> {
        let id = Uuid::new_v4().to_string();
        let mut stored = entry.clone();
        stored.id = id.clone();

        // Assign wall-clock if the caller left it at zero.
        if stored.created_at == 0.0 {
            stored.created_at = now_secs();
        }

        self.entries.write().await.push(stored);
        Ok(id)
    }

    async fn search(&self, query: &str, limit: usize) -> Result<Vec<LongTermEntry>, RunnerError> {
        let query_lower = query.to_lowercase();
        let guard = self.entries.read().await;

        let mut matches: Vec<LongTermEntry> = guard
            .iter()
            .filter(|e| e.content.to_lowercase().contains(&query_lower))
            .cloned()
            .collect();

        // Newest first.
        matches.sort_by(|a, b| b.created_at.partial_cmp(&a.created_at).unwrap_or(std::cmp::Ordering::Equal));
        matches.truncate(limit);
        Ok(matches)
    }

    async fn get(&self, id: &str) -> Result<Option<LongTermEntry>, RunnerError> {
        let guard = self.entries.read().await;
        Ok(guard.iter().find(|e| e.id == id).cloned())
    }

    async fn delete(&self, id: &str) -> Result<(), RunnerError> {
        let mut guard = self.entries.write().await;
        guard.retain(|e| e.id != id);
        Ok(())
    }

    async fn list_recent(&self, limit: usize) -> Result<Vec<LongTermEntry>, RunnerError> {
        let guard = self.entries.read().await;
        let mut all: Vec<LongTermEntry> = guard.clone();

        // Newest first.
        all.sort_by(|a, b| b.created_at.partial_cmp(&a.created_at).unwrap_or(std::cmp::Ordering::Equal));
        all.truncate(limit);
        Ok(all)
    }
}

/// Current wall-clock as fractional seconds (UTC).
fn now_secs() -> f64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64()
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use super::*;

    fn entry(content: &str) -> LongTermEntry {
        LongTermEntry {
            id: String::new(),
            entry_type: "learning".into(),
            content: content.into(),
            tags: vec![],
            session_id: None,
            created_at: 0.0,
            metadata: HashMap::new(),
        }
    }

    #[tokio::test]
    async fn save_assigns_uuid() {
        let backend = InMemoryBackend::new();
        let id = backend.save(&entry("hello")).await.unwrap();
        assert!(!id.is_empty());
        // UUID v4 format: 8-4-4-4-12 hex chars
        assert_eq!(id.len(), 36);
    }

    #[tokio::test]
    async fn search_is_case_insensitive() {
        let backend = InMemoryBackend::new();
        backend.save(&entry("UPPERCASE content")).await.unwrap();
        backend.save(&entry("lowercase content")).await.unwrap();
        backend.save(&entry("no match here")).await.unwrap();

        let results = backend.search("content", 10).await.unwrap();
        assert_eq!(results.len(), 2);
    }

    #[tokio::test]
    async fn search_respects_limit() {
        let backend = InMemoryBackend::new();
        for i in 0..5 {
            backend
                .save(&entry(&format!("entry {i}")))
                .await
                .unwrap();
        }

        let results = backend.search("entry", 2).await.unwrap();
        assert_eq!(results.len(), 2);
    }

    #[tokio::test]
    async fn get_returns_none_for_missing() {
        let backend = InMemoryBackend::new();
        assert!(backend.get("nonexistent").await.unwrap().is_none());
    }

    #[tokio::test]
    async fn delete_is_idempotent() {
        let backend = InMemoryBackend::new();
        // Deleting a missing ID is not an error.
        backend.delete("nonexistent").await.unwrap();
    }

    #[tokio::test]
    async fn list_recent_respects_limit() {
        let backend = InMemoryBackend::new();
        for i in 0..5 {
            backend
                .save(&entry(&format!("entry {i}")))
                .await
                .unwrap();
        }

        let recent = backend.list_recent(3).await.unwrap();
        assert_eq!(recent.len(), 3);
    }
}
